File: app/src/main/java/com/linwoain/storytelling/MainActivity.java
package com.linwoain.storytelling;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ListView;
import cn.bmob.v3.Bmob;
import cn.bmob.v3.update.BmobUpdateAgent;
import com.jakewharton.rxbinding.widget.RxAdapterView;
import com.koushikdutta.async.future.FutureCallback;
import com.koushikdutta.ion.Ion;
import com.linwoain.storytelling.adapter.BookAdapter;
import com.linwoain.storytelling.bean.BookInfo;
import com.linwoain.storytelling.bean.ChapterBean;
import com.linwoain.storytelling.bus.Progress;
import com.linwoain.storytelling.bus.RxBus;
import com.linwoain.storytelling.common.CommonActivity;
import com.linwoain.storytelling.config.Constant;
import com.linwoain.util.CacheUtil;
import com.linwoain.util.LLogUtils;
import com.linwoain.util.Once;
import com.umeng.analytics.MobclickAgent;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;

public class MainActivity extends CommonActivity {

  ListView listView;
  List<BookInfo> books = new ArrayList<>();
  BookAdapter adapter;
  Subscription register;
  private String flowerUrl =
      "http://42.121.125.229:8080/audible-book/service/audioBooksV2/getBookChaptersByPage?market=k-app360&refer=SearchResultBookList&dir=DESC&token=QUi2Cp%2BU09ITUXjudFZ2CtiCKUWfP1%2FT5KepiuHKXM77WDCYJwpQNoxu37pltHhH&pageSize=200&bookId=190518&imsi=460072521701419&ver=3.9.1&appKey=audibleBook";
  private String huaxuyinUrl =
      "http://42.121.125.229:8080/audible-book/service/audioBooksV2/getBookChaptersByPage?market=k-app360&refer=SearchResultBookList&dir=ASC&token=QUi2Cp%2BU09ITUXjudFZ2CtiCKUWfP1%2FT5KepiuHKXM77WDCYJwpQNoxu37pltHhH&pageSize=200&bookId=212947&imsi=460072521701419&ver=3.9.1&appKey=audibleBook";

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    LLogUtils.setTag("book");
    initUmeng();
    initView();
    getData();
    registerEventBus();
  }

  private void initUmeng() {
    MobclickAgent.setScenarioType(this, MobclickAgent.EScenarioType.E_UM_NORMAL);
    Bmob.initialize(this,"df810df15d0eabd12c6550465f6fcfc4");
    Once.execute("update", new Once.OnceCallback() {
      @Override public void onOnce() {
        BmobUpdateAgent.initAppVersion(MainActivity.this);
      }
    });
    BmobUpdateAgent.update(this);
  }

  private void initView() {
    Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
    listView = (ListView) findViewById(R.id.listview);
    setSupportActionBar(toolbar);
    RxAdapterView.itemClicks(listView)
        .throttleFirst(500, TimeUnit.MILLISECONDS)
        .subscribeOn(AndroidSchedulers.mainThread())
        .subscribe(new Action1<Integer>() {
          @Override public void call(Integer integer) {
            BookInfo bean = books.get(integer);
            startActivity(
                new Intent(MainActivity.this, MuluActivity.class).putExtra(Constant.BEAN, bean));
          }
        });
  }

  @Override public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu_main, menu);
    return true;
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    int id = item.getItemId();

    if (id == R.id.action_settings) {
      return true;
    }

    return super.onOptionsItemSelected(item);
  }

  private void getData() {
    Ion.with(this).load(flowerUrl).as(BookInfo.class).setCallback(new FutureCallback<BookInfo>() {
      @Override public void onCompleted(Exception e, BookInfo result) {
        if (e == null) {
          result.setName("花千骨");
          result.setUrl(flowerUrl);
          result.setBookId(190518);
          books.add(result);
          refreshProgress();
          setData();
        } else {
          e.printStackTrace();
        }
      }
    });
    Ion.with(this).load(huaxuyinUrl).as(BookInfo.class).setCallback(new FutureCallback<BookInfo>() {
      @Override public void onCompleted(Exception e, BookInfo result) {
        if (e == null) {
          result.setName("华胥引");
          result.setUrl(huaxuyinUrl);
          result.setBookId(212947);
          books.add(result);
          refreshProgress();
          setData();
        } else {
          e.printStackTrace();
        }
      }
    });
  }

  private void registerEventBus() {

    register = RxBus.getDefault().toObserverable(Progress.class).subscribe(new Action1<Progress>() {
      @Override public void call(Progress bean) {
        for (BookInfo book : books) {
          if (book.getBookId() == bean.getBookId()) {
            book.setTitle(bean.getTitle());
            setData();
          }
        }
      }
    });
  }

  @Override protected void onDestroy() {
    if (register.isUnsubscribed()) {
      register.unsubscribe();
    }
    super.onDestroy();
  }

  @Override protected void onRestart() {
    super.onRestart();
    refreshProgress();
  }

  private void refreshProgress() {
    for (BookInfo book : books) {
      ChapterBean bean = (ChapterBean) CacheUtil.getBaseBean(Constant.MULU + book.getBookId());
      if (bean != null) {

        book.setTitle(bean.getTitle());
        setData();
      }
    }
  }

  private void setData() {
    if (adapter == null) {
      adapter = new BookAdapter(this, books);
      listView.setAdapter(adapter);
    } else {
      adapter.notifyDataSetChanged();
    }
  }
}

File: app/src/main/java/com/linwoain/storytelling/MainActivity.kt
package com.linwoain.storytelling

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.app.AppCompatDelegate
import android.view.Menu
import android.view.MenuItem
import com.linwoain.storytelling.adapter.BookAdapter
import com.linwoain.storytelling.bean.Novel
import com.linwoain.storytelling.bean.NovelWrapper
import com.linwoain.storytelling.bus.Progress
import com.linwoain.storytelling.config.Constant
import com.linwoain.storytelling.utils.process
import com.linwoain.util.CacheUtil
import com.linwoain.util.GsonUtil
import com.orhanobut.logger.Logger
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import org.litepal.crud.DataSupport
import java.util.*


class MainActivity : AppCompatActivity() {

    internal var books: ArrayList<Novel> = ArrayList()

    internal var adapter: BookAdapter? = null

    companion object {
        private val NOVEL_LIST_URL = "https://bwg.linwoain.com/novel"
        private val BUNDLE_BOOKS = "BUNDLE_BOOKS"

    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        registerEventBus()
        initView()
        val tempBooks = savedInstanceState?.getSerializable(BUNDLE_BOOKS)
        if (tempBooks != null) {
            books.addAll(tempBooks as ArrayList<Novel>)
        }
        if (books.size == 0) {
            getData()
        } else {
            setData()
        }

    }


    private fun initView() {
        setSupportActionBar(toolbar)
        listview.setOnItemClickListener { _, _, position, _ -> startActivity<MuluActivity>(Constant.BEAN to books[position]) }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_about -> startActivity<AboutActivity>()
            R.id.action_night -> {
                val nightMode = CacheUtil.getBoolean(Constant.NIGHT_MODE)
                val builder = StringBuilder(nightMode.toString())

                if (!nightMode) {//非夜间模式时，设置为夜间模式
                    delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                    CacheUtil.save(Constant.NIGHT_MODE, true)
                    builder.append(" -> true")
                    recreate()
                } else {
                    delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                    CacheUtil.save(Constant.NIGHT_MODE, false)
                    builder.append(" -> false")

                    recreate()
                }
                Logger.d(builder.toString())
            }
            else -> toast("应该不可能发生的事情发生了😯")
        }

        return true
    }

    private fun getData() {

        val data = getDataFromDB()
        if (data.isNotEmpty()) {
            Logger.d("book data from db")
            books.addAll(data)
            setData()
            return
        }
        getDataFromNet()


    }

    private fun getDataFromNet() {
        process(NOVEL_LIST_URL) {
            val wrapper = GsonUtil.get(it, NovelWrapper::class.java)
            books.addAll(wrapper.novels)
            setDataToDB()
            setData()
        }
    }

    private fun setDataToDB() {
        books.forEach {
            it.save()
        }
    }

    private fun getDataFromDB(): List<Novel> {
        return DataSupport.findAll(Novel::class.java)

    }

    private fun registerEventBus() {
        EventBus.getDefault().register(this)
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    fun event(progress: Progress) {
        books.forEach {
            if (it.bookId == progress.bookId) {
                it.title = progress.title
                setData()
            }
        }
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }


    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        if (books.size == 2) {
            outState?.putSerializable(BUNDLE_BOOKS, books)
        }
    }


    private fun setData() {
        if (adapter == null) {
            adapter = BookAdapter(this, books)
            listview.adapter = adapter
        } else {
            adapter?.notifyDataSetChanged()
        }
    }
}
